File: tictactoe_flutter/lib/game_logic.dart
class GameLogic {
  List<List<String>> board = List.generate(3, (_) => List.filled(3, "", growable: false));
  String currentPlayer = "X";

  bool makeMove(int row, int col) {
    if (board[row][col].isEmpty) {
      board[row][col] = currentPlayer;
      if (checkWin() || isDraw()) {
        return true;
      }
      _switchPlayer();
    }
    return false;
  }

  void _switchPlayer() {
    currentPlayer = (currentPlayer == "X") ? "O" : "X";
  }

  bool checkWin() {
    for (int i = 0; i < 3; i++) {
      if (_isSame(board[i][0], board[i][1], board[i][2])) return true;
      if (_isSame(board[0][i], board[1][i], board[2][i])) return true;
    }
    if (_isSame(board[0][0], board[1][1], board[2][2])) return true;
    if (_isSame(board[0][2], board[1][1], board[2][0])) return true;
    return false;
  }

  bool isDraw() {
    for (var row in board) {
      if (row.contains("")) return false;
    }
    return !checkWin();
  }

  void resetGame() {
    board = List.generate(3, (_) => List.filled(3, "", growable: false));
    currentPlayer = "X";
  }

  bool _isSame(String a, String b, String c) {
    return a.isNotEmpty && a == b && b == c;
  }
}

File: tictactoe_android/app/src/main/java/com/robin/bachelor/thesis/tictactoe_android/GameLogic.kt
package com.robin.bachelor.thesis.tictactoe_android

class GameLogic {
  var board: Array<Array<String>> = Array(3) { Array(3) { "" } }
  var currentPlayer: String = "X"

  fun makeMove(row: Int, col: Int): Boolean {
    if (board[row][col].isEmpty()) {
      board[row][col] = currentPlayer
      if (checkWin() || isDraw()) return true
      switchPlayer()
    }
    return false
  }

  private fun switchPlayer() {
    currentPlayer = if (currentPlayer == "X") "O" else "X"
  }

  fun checkWin(): Boolean {
    for (i in 0..2) {
      if (isSame(board[i][0], board[i][1], board[i][2])) return true
      if (isSame(board[0][i], board[1][i], board[2][i])) return true
    }
    if (isSame(board[0][0], board[1][1], board[2][2])) return true
    if (isSame(board[0][2], board[1][1], board[2][0])) return true
    return false
  }

  fun isDraw(): Boolean {
    return board.all { row -> row.all { it.isNotEmpty() } } && !checkWin()
  }

  fun resetGame() {
    board = Array(3) { Array(3) { "" } }
    currentPlayer = "X"
  }

  private fun isSame(a: String, b: String, c: String): Boolean {
    return a.isNotEmpty() && a == b && b == c
  }
}
